### PR TITLE
rgw-admin: new commands to control placement

### PR DESCRIPTION
**Pull request by yehudasa**

### CODE REVIEW

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -3450,6 +3539,79 @@ int main(int argc, char **argv)
...
+        if (placement_id.empty()) {
+          cerr << "ERROR: --placement-id not specified" << std::endl;
+          return EINVAL;
+        }
+	RGWRealm realm(realm_id, realm_name);
+	int ret = realm.init(g_ceph_context, store);
+	if (ret < 0) {
+	  cerr << "failed to init realm: " << cpp_strerror(-ret) << std::endl;
+	  return -ret;
+	}
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc]: the realm looks unused here

File: src/rgw/rgw_admin.cc [hunk 2]
```diff
@@ -3450,6 +3539,79 @@ int main(int argc, char **argv)
...
+
+        if (opt_cmd == OPT_ZONEGROUP_PLACEMENT_ADD) {
+          RGWZoneGroupPlacementTarget target;
+          target.name = placement_id;
+          for (auto& t : tags) {
+            target.tags.insert(t);
+          }
+          zonegroup.placement_targets[placement_id] = target;
+        } else if (opt_cmd == OPT_ZONEGROUP_PLACEMENT_MODIFY) {
+          RGWZoneGroupPlacementTarget& target = zonegroup.placement_targets[placement_id];
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc]: maybe use `zonegroup.placement_targets.find(placement_id)` and return an error if it didn't exist?

File: src/rgw/rgw_admin.cc [hunk 3]
```diff
@@ -213,7 +221,16 @@ void _usage()
   cout << "   --source-zone             specify the source zone (for data sync)\n";
   cout << "   --default                 set entity (realm, zonegroup, zone) as default\n";
   cout << "   --read-only               set zone as read-only (when adding to zonegroup)\n";
+  cout << "   --placement-id            placement id form zonegroup placement commands\n";
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc]: `form` -> `for`

File: src/rgw/rgw_admin.cc [hunk 4]
```diff
@@ -3900,6 +4062,58 @@ int main(int argc, char **argv)
...
+	  return -ret;
+	}
+
+        if (opt_cmd == OPT_ZONE_PLACEMENT_ADD ||
+            opt_cmd == OPT_ZONE_PLACEMENT_MODIFY) {
+          RGWZonePlacementInfo& info = zone.placement_pools[placement_id];
+
+          info.index_pool = index_pool;
+          info.data_pool = data_pool;
+          info.data_extra_pool = data_extra_pool;
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc]: should we fail if index_pool or data_pool are empty?

Comment (yehudasa) [inline src/rgw/rgw_admin.cc]: ok

File: src/rgw/rgw_admin.cc [hunk 5]
```diff
@@ -3900,6 +4056,63 @@ int main(int argc, char **argv)
...
+        if (ret < 0) {
+	  cerr << "failed to init zone: " << cpp_strerror(-ret) << std::endl;
+	  return -ret;
+	}
+
+        if (opt_cmd == OPT_ZONE_PLACEMENT_ADD ||
+            opt_cmd == OPT_ZONE_PLACEMENT_MODIFY) {
+          RGWZonePlacementInfo& info = zone.placement_pools[placement_id];
+
+          if (index_pool.empty() || data_pool.empty) {
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc]: `data_pool.empty` needs some `()`

**Pull request merged**